flux-start: add embedded server

garlick opened the pull request:
This adds a convenience function for setting up an embedded server that listens on a `local://` socket and routes messages between connected clients and a server side `flux_t` handle.  It's a bit of an abomination if you think about it too hard, but it does let arbitrary programs offer "services" in the usual way by registering message handlers, and clients to connect to it and use RPCs the way they would do with flux, so pretty convenient.

Maybe this will come in handy if the shell needs to offer a standalone service to applications, especially if that application is flux?

A server is then embedded in `flux-start` for the purpose of enabling system tests that restart brokers.  In this PR, the only service method just lists broker pids.  I thought maybe getting this much in with tests was maybe a good place to cut this PR and start the next one?

grondo commented:
This is pretty neat. My mind is starting to explore all the neat things for which this could be used.

The flux-shell application is a good example. Does this open up the ability for programs to route messages as well?

How do you derive the URI for one of these connections? Could there be a special URI to attach to a named process with `flux_open()` for convenience (e..g `flux_open ("local:flux-start", 0)`)? (eh, that is probably getting a little off-topic here)

garlick commented:
> Does this open up the ability for programs to route messages as well?

librouter has other abstractions that make that relatively straightforward (e.g. see `flux-proxy` or `connector-local`), although nothing prevents a program that embeds one of these servers from routing messages too, at least that I can think of right now.  There isn't actually that much that is new here - for the most part it is just repackaging a few things that were already implemented.

> How do you derive the URI for one of these connections? Could there be a special URI to attach to a named process with flux_open() for convenience (e..g flux_open ("local:flux-start", 0))? (eh, that is probably getting a little off-topic here)

Heh, do you mean like have a special URI path that instructs flux_open to internally substitute the final component of FLUX_URI?   That's kind of a neat idea that would save a bit of code.  Let me think about that one.

grondo commented:
> Edit: flux-start locates the socket in rundir, so if FLUX_URI is set to local:///tmp/flux-gmnMYf/local-0 then flux-start can be reached at local:///tmp/flux-gmnMYf/start. But that's just the convention I implemented for this use case - kind of arbitrary.

Well there is some precedent for this (I'm thinking of something like `tmux`, which places all sockets under a common path for discovery). What I was thinking we could allow is that the server-side could provide a name (e.g. "flux-start" or just "start", or for the shell "flux-shell") when opening the socket. Then a client could discover available sockets and connect by name (again thinking of the `tmux` example). It is difficult to imagine this would be very useful, since I can't think of other non-broker processes that would want to open one of these sockets, it was more of a spur of the moment idea. At the very least each client would not be required to substitute paths in FLUX_URI by hand.

Sorry if the idea got you a little off track.

garlick commented:
I'm not sure that would work for the shell, since if there are multiple jobs running under the same broker, the rundir directory is not unique.  Also the shell may not have write permission there to create a socket (as a guest).  There might be other approaches for "socket discovery" though that could make a flux ecosystem easier to navigate, but maybe that's a topic for another day?

garlick commented:
Just realized I had not documented another caveat - that the server `flux_t` handle requires a reactor to make progress.  Added header comments and updated commit message.

grondo commented:
> Also the shell may not have write permission there to create a socket (as a guest). There might be other approaches for "socket discovery" though that could make a flux ecosystem easier to navigate, but maybe that's a topic for another day?

Yeah, duh that is a problem. 🤦 Yeah, we'll just kick the can down the road awhile longer. It is not likely we'll add this capability to the shell anytime soon anyway, and no use discussing this for test-only flux-start use

garlick commented:
> At the very least each client would not be required to substitute paths in FLUX_URI by hand.

I forced a push a change to set a new environment variable `FLUX_START_URI` to relieve clients of the burden of substituting strings.  I'm not a great fan of proliferating Flux environment variables, but for testing only maybe this is OK.

grondo commented:
> but for testing only maybe this is OK.

`flux start` isn't just for testing anymore (works well for a single node, single user instance. Though I think the impact starting the internal flux-start server and resulting `FLUX_START_URI` is minimal, perhaps the server should be behind a `--test-*` option?

Also, `flux-start` should probably clear `FLUX_START_URI` from the environment at startup so it doesn't get passed through to subinstances. Again probably low risk, but it might be confusing if your subinstance connects to a flux-start from a different instance.

garlick commented:
As discussed at coffee, I don't think it's necessary to provide an option to enable/disable this as the service is low overhead and seems unlikely to get in the way.  We could easily add it later if needed.

We can't sanitize FLUX_START_URI in the broker because we need it to be passed through when a test program is run as the initial program.

Should we sanitize it in the job shell, e.g.

```diff
diff --git a/src/shell/task.c b/src/shell/task.c
index 5d1912db7..a90123d01 100644
--- a/src/shell/task.c
+++ b/src/shell/task.c
@@ -150,6 +150,9 @@ struct shell_task *shell_task_create (struct shell_info *info,
                               getenv ("FLUX_KVS_NAMESPACE")) < 0)
             goto error;
     }
+    /* Sanitize FLUX variables that should not propagate to jobs.
+     */
+    flux_cmd_unsetenv (task->cmd, "FLUX_START_URI");
     return task;
 error:
     shell_task_destroy (task);
```

It feels slightly wrong, like we're on a slippery slope to complex environment propagation rules.

Maybe it's a better idea to revert the change that added the env var?

grondo commented:
> It feels slightly wrong, like we're on a slippery slope to complex environment propagation rules.
> Maybe it's a better idea to revert the change that added the env var?

I think you had convinced me that propagation of `FLUX_START_URI` to all jobs and child instances was not a big deal at this point. I say your current approach is fine for now. Improvements, if necessary, could always be done in the future.

garlick commented:
OK, then let's leave it in and press forward.  Thanks for the discussion!

I restarted CI - one builder got stuck on that shell input sharness test, which I think is a known one.

grondo reviewed (approved):
As discussed, LGTM!

chu11 commented on src/common/libflux/message.h:
```diff
@@ -191,7 +191,7 @@ int flux_msg_set_flags (flux_msg_t *msg, uint8_t flags);
  * flux_msg_set_string() accepts a NULL 's' (no payload).
  * flux_msg_get_string() will set 's' to NULL if there is no payload
- * N.B. the raw paylaod includes C string \0 terminator.
+ * N.B. the raw payload includes C string \0 terminator.
```
commit message typo? I see only one comment misspelling, not two :-)

garlick replied:
Huh, weird.  I could have sworn there were two.  Oh, found the same misspelled word in event.h.  I'll tack that on and fix the commit message.

garlick commented:
OK, forced a push with a slightly expanded set of spelling corrections and a more vague header :-) Thanks for that.  I'll set MWP.
